source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+    return std::make_tuple(
+        std::unique_ptr<Value>(new ScalarValue(result_in_raw_word)),
+        std::vector<uint32_t>());
+
+  } else {
+    assert(
+        AllInOperandValuesAreConstVectorsInSameLength(inst, id_to_const_val_) &&
+        "All the operands must be vector constants with same number of "
+        "components");
+    (void)static_cast<bool (*)(
+        ir::Instruction*,
+        const std::unordered_map<uint32_t, std::unique_ptr<Value>>&)>(
+        AllInOperandValuesAreConstVectorsInSameLength);
+
+    CompositeValue* first_operand_value = iter->second->AsComposite();
+    uint32_t num_dims =
+        static_cast<uint32_t>(first_operand_value->nested.size());
+    // Get the value in raw words for all the operands in each dimension. The
+    // result from each dimension should be aggregated to get the final result
+    // vector.
+    auto result = std::unique_ptr<Value>(new CompositeValue);
+    std::vector<uint32_t> new_component_ids;
+    for (uint32_t d = 0; d < num_dims; d++) {
+      std::vector<uint32_t> one_dimension_operand_values_in_raw_words;
+      one_dimension_operand_values_in_raw_words.reserve(num_operands -
+                                                        spec_opcode_count);
+      one_dimension_operand_values_in_raw_words.push_back(
+          id_to_const_val_[first_operand_value->nested[d]]->AsScalar()->word);
+      for (uint32_t i = operand_index + 1; i < num_operands; i++) {
+        iter = id_to_const_val_.find(inst->GetSingleWordInOperand(i));
+        if (iter == id_to_const_val_.end()) {
+          return std::make_tuple(nullptr, std::vector<uint32_t>());
+        }
+        one_dimension_operand_values_in_raw_words.push_back(
+            id_to_const_val_[iter->second->AsComposite()->nested[d]]
+                ->AsScalar()
+                ->word);
+      }
+      uint32_t one_dimension_result =
+          Operate(static_cast<SpvOp>(inst->GetSingleWordInOperand(0)),
+                  one_dimension_operand_values_in_raw_words);
+      auto one_dimension_result_val =
+          std::unique_ptr<Value>(new ScalarValue(one_dimension_result));
+      uint32_t new_id = GetNewId();
+      id_to_const_val_[new_id] = std::move(one_dimension_result_val);
+      result->AsComposite()->nested.push_back(new_id);
+      new_component_ids.emplace_back(new_id);
+    }
+    return std::make_tuple(std::move(result), new_component_ids);
+  }
+}
+
+ir::Module::inst_iterator FoldSpecConstantOpAndCompositePass::InsertConstantAt(
+    const analysis::Type* type, uint32_t result_id,
+    ir::Module::inst_iterator inst_iter) {
+  std::unique_ptr<ir::Instruction> new_inst;
+  uint32_t type_id = type_mgr_->GetId(type);
+  auto iter = id_to_const_val_.find(result_id);
+  assert(iter != id_to_const_val_.end() &&
+         "The value for the new constant should be added to id_to_const_val_ "
+         "map before creating and inserting its instruction");
+
+  if (iter->second->AsScalar()) {
```

Comment:
It seems to me that we can add a new method to the `Value` class hierarchy, like `ToInst(type_id, result_id)`, to turn a value into a `std::unique_ptr<Inst>`. So we can make the code looks a little bit more organized.